datumaro/plugins/synthetic_data/image_generator.py

```diff
@@ -0,0 +1,268 @@
+            for multiplier in INTERVAL_MULTIPLIERS:
+                modified_weights = BASE_WEIGHTS.copy()
+                modified_weights[weight_index] += multiplier * WEIGHT_INTERVAL
+                weight_vectors.append(modified_weights)
+        weights = np.array(weight_vectors)
+        return weights
+
+    @classmethod
+    def _download_colorization_model(cls, save_dir: str) -> None:
+        prototxt_file_name = cls._MODEL_PROTO_FILENAME
+        caffemodel_file_name = cls._MODEL_WEIGHTS_FILENAME
+        hull_file_name = cls._HULL_PTS_FILE_NAME
+
+        proto_path = osp.join(save_dir, prototxt_file_name)
+        model_path = osp.join(save_dir, caffemodel_file_name)
+        hull_path = osp.join(save_dir, hull_file_name)
+        if not (
+            osp.exists(proto_path) and osp.exists(model_path) and osp.exists(hull_path)
+        ) and not os.access(save_dir, os.W_OK):
+            raise ValueError(
+                "Please provide a path to a colorization model directory or "
+                "a path to a writable directory to download the model"
+            )
+
+        for url, filename, size, md5_checksum in [
+            (
+                f"https://raw.githubusercontent.com/richzhang/colorization/a1642d6ac6fc80fe08885edba34c166da09465f6/colorization/models/{prototxt_file_name}",
+                prototxt_file_name,
+                9945,
+                "7229f469d24645a7f1e3c47c67e7bd15",
+            ),
+            (
+                f"http://eecs.berkeley.edu/~rich.zhang/projects/2016_colorization/files/demo_v2/{caffemodel_file_name}",
+                caffemodel_file_name,
+                128946764,
+                "6ef9d30cefd880baabeb5e1847fb20be",
+            ),
+            (
+                f"https://raw.githubusercontent.com/richzhang/colorization/a1642d6ac6fc80fe08885edba34c166da09465f6/colorization/resources/{hull_file_name}",
+                hull_file_name,
+                5088,
+                "05c2729d850b5e4143b6fe53326066b5",
+            ),
+        ]:
+            save_path = osp.join(save_dir, filename)
+            if osp.exists(save_path):
+                continue
+
+            log.info("Downloading the '%s' file to '%s'", filename, save_dir)
+            try:
+                cls._download_file(
+                    url, save_path, expected_size=size, expected_checksum=md5_checksum
+                )
+            except Exception as e:
+                raise Exception(f"Failed to download the '{filename}' file: {str(e)}")
+
+    @staticmethod
+    @scoped
+    def _download_file(
+        url: str, output_path: str, *, timeout: int = 60, expected_size: int, expected_checksum: str
+    ) -> None:
+        BLOCK_SIZE = 2**20
+
+        response = requests.get(url, timeout=timeout, stream=True)
+        response.raise_for_status()
+
+        checksum_counter = hashlib.md5()  # nosec B303 - not a security context
+        actual_size = 0
+
+        on_error_do(os.unlink, output_path)
```

Comment:
This will crash if the error occurs before the file is created (i.e., in `open`).